src/Glpi/Controller/Security/MFAController.php

```diff
@@ -0,0 +1,155 @@
+    )]
+    #[SecurityStrategy(Firewall::STRATEGY_NO_CHECK)]
+    public function setup(Request $request): Response
+    {
+        global $CFG_GLPI;
+        if (!isset($_SESSION['mfa_pre_auth'])) {
+            return new RedirectResponse($CFG_GLPI['root_doc'] . '/front/login.php');
+        }
+        return new StreamedResponse(static function () {
+            $totp = new TOTPManager();
+            $totp->showTOTPSetupForm((int) $_SESSION['mfa_pre_auth']['user']['id']);
+        });
+    }
+
+    #[Route(
+        path: "/MFA/Prompt",
+        name: "mfa_prompt",
+        methods: ['GET']
+    )]
+    #[SecurityStrategy(Firewall::STRATEGY_NO_CHECK)]
+    public function prompt(Request $request): Response
+    {
+        $totp = new TOTPManager();
+        return new StreamedResponse(static function () use ($totp) {
+            $totp->showTOTPPrompt();
+        });
+    }
+
+    #[Route(
+        path: "/MFA/Verify",
+        name: "mfa_verify_totp",
+        methods: ['POST']
+    )]
+    #[SecurityStrategy(Firewall::STRATEGY_NO_CHECK)]
+    public function verify(Request $request): Response
+    {
+        global $CFG_GLPI;
+
+        $from_login = isset($_SESSION['mfa_pre_auth']);
+        $users_id = $from_login ? (int) $_SESSION['mfa_pre_auth']['user']['id'] : (int) Session::getLoginUserID();
+        if (!$users_id) {
+            return new RedirectResponse($request->getBasePath() . '/MFA/Prompt');
+        }
+        $totp = new TOTPManager();
+        $backup_code = $request->request->get('backup_code');
+        $totp_code = $request->get('totp_code');
+        if (is_array($totp_code)) {
+            $totp_code = implode('', $totp_code);
+        }
+        $secret = $request->request->get('secret');
+        $algorithm = null;
+
+        if (
+            !(
+                (isset($backup_code) && $totp->verifyBackupCodeForUser($backup_code, $users_id))
+                || (isset($totp_code, $secret) && ($algorithm = $totp->verifyCodeForSecret($totp_code, $secret)))
+                || (isset($totp_code) && !isset($secret) && $totp->verifyCodeForUser($totp_code, $users_id))
+            )
+        ) {
+            // Verification failure
+            if (isset($backup_code)) {
+                throw new AuthenticationFailedException(authentication_errors: [__('Invalid backup code')]);
+            } else {
+                throw new AuthenticationFailedException(authentication_errors: [__('Invalid TOTP code')]);
+            }
+        }
+
+        if (
+            isset($secret)
+            && !(Session::validateIDOR($request->request->all())
+                && $totp->setSecretForUser($users_id, $request->request->getString('secret'), $algorithm))
+        ) {
+            Session::addMessageAfterRedirect(__s('Invalid code'), false, ERROR);
+            return new RedirectResponse($from_login ? ($CFG_GLPI['root_doc'] . '/MFA/Prompt') : Html::getBackUrl());
+        }
+
+        $_SESSION['mfa_success'] = true;
```

Comment:
Is this unset somewhere? We to be sure that a user cannot reuse a successful MFA login in a new login attempt.